exonum_build/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// spell-checker:ignore walkdir, subfolders, submodules
+
+//! This crate simplifies writing build.rs for exonum and exonum services.
+
+extern crate protoc_rust;
+extern crate walkdir;
+
+use protoc_rust::Customize;
+use walkdir::WalkDir;
+
+use std::{env, fs::File, io::Write, path::Path};
+
+/// Find all .proto files in `path` and subfolders.
+fn get_proto_files<P: AsRef<Path>>(path: P) -> Vec<String> {
+    WalkDir::new(path)
+        .into_iter()
+        .filter_map(|e| {
+            let e = e.ok()?;
+            if e.path().extension()?.to_str() == Some("proto") {
+                Some(e.path().to_str()?.to_owned())
+            } else {
+                None
+            }
+        }).collect()
+}
+
+/// Workaround for https://github.com/stepancheg/rust-protobuf/issues/324
+/// It is impossible to `include!` .rs files generated by rust-protobuf
+/// so we generate piece of `mod.rs` which includes generated files as public submodules.
+///
+/// tests.proto .rs file will be included with `#[cfg(test)]`.
+fn generate_mod_rs(out_dir: &str, proto_files: &[String], mod_file: &str) {
+    let mod_file_content = {
+        proto_files
+            .iter()
+            .map(|f| {
+                let mod_name = Path::new(f)
+                    .file_stem()
+                    .unwrap()
+                    .to_str()
+                    .expect("proto file name is not &str");
+                if mod_name == "tests" {
+                    format!("#[cfg(test)]\npub mod {};\n", mod_name)
+                } else {
+                    format!("pub mod {};\n", mod_name)
+                }
+            }).collect::<String>()
+    };
+    let dest_path = Path::new(&out_dir).join(mod_file);
+    let mut file = File::create(dest_path).expect("Unable to create output file");
+    file.write_all(mod_file_content.as_bytes())
+        .expect("Unable to write data to file");
+}
+
+/// Generate .rs files from .proto files.
+///
+/// `protoc` executable from protobuf should be in `$PATH`
+///
+/// # Examples
+///
+/// In `build.rs`
+/// ```no_run
+/// extern crate exonum_build;
+///
+/// use exonum_build::protobuf_generate;
+///
+/// // Includes usually should contain input_dir.
+/// protobuf_generate("src/proto", &["src/proto"], "example_mod.rs")
+/// ```
+/// After successful run `$OUT_DIR` will contain \*.rs for each \*.proto file in
+/// "src/proto/\*\*/" and example_mod.rs which will include all generated .rs files
+/// as submodules.
+///
+/// To use generated protobuf structs.
+///
+/// In `src/proto/mod.rs`
+/// ```ignore
+/// extern crate exonum;
+///
+/// include!(concat!(env!("OUT_DIR"), "/example_mod.rs"));
+///
+/// // If you use types from exonum .proto files.
+/// use exonum::encoding::protobuf::*;
+/// ```
+pub fn protobuf_generate(input_dir: &str, includes: &[&str], mod_file: &str) {
```

Comment:
```suggestion
pub fn protobuf_generate<P: AsRef<Path>>(input_dir: P, includes: &[&Path], mod_file_name: &str) {
```

Or even `impl Iterator` instead of slice

Reply:
Ok, but the reason it is like this is that protoc crate uses &str everywhere.
Interface may look a little bit better but it will contain same conversions to String.

Reply:
@aleksuss Thanks for suggesting correct path.
You solution is close but does not work perfectly, it requires you to write
`protobuf_generate(&"src/proto", &["src/proto"], &"example_mod.rs")`
I found correct one to be `impl IntoIterator<Item=impl AsRef<Path>` this way it does not tie together types of first argument of the function(input_dir) and second one(includes).

Also this solution has another flaw - input_dir and mod_file_name types now tied together and it will be type error to use different types which are `AsRef<Path>` there, so if you want to be really generic you have to put another template parameter for mod_file_name.